src/typechecker/FStar.TypeChecker.Rel.fs

```diff
@@ -3860,22 +3933,137 @@ let resolve_implicits' env must_total forcelax g =
                until_fixpoint (g'.implicits@out, true) tl in
   {g with implicits=until_fixpoint ([], false) g.implicits}
 
-let resolve_implicits env g = resolve_implicits' env (not env.phase1 && not env.lax)  false g
+let resolve_implicits env g =
+    if Env.debug env <| Options.Other "ResolveImplicitsHook"
+    then BU.print1 "//////////////////////////ResolveImplicitsHook: resolve_implicits////////////\n\
+                    guard = %s\n"
+                    (guard_to_string env g);
+  resolve_implicits' env (not env.phase1 && not env.lax) false g
+
 let resolve_implicits_tac env g = resolve_implicits' env false true  g
 
 let force_trivial_guard env g =
-    let g = solve_deferred_constraints env g |> resolve_implicits env in
+    if Env.debug env <| Options.Other "ResolveImplicitsHook"
+    then BU.print1 "//////////////////////////ResolveImplicitsHook: force_trivial_guard////////////\n\
+                    guard = %s\n"
+                    (guard_to_string env g);
+
+    let solve_goals_with_tac (imps:implicits) (tac:sigelt) =
```

Comment:
Would be nice to split this whole chunk into a separate function, so `force_trivial_guard` remains more readable, but we can do that in a later cleanup.

Reply:
That's a good suggestion. I can do that before merging.